src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "AsyncTransferFacilitator.h"
+
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace bdx {
+
+CHIP_ERROR AsyncTransferFacilitator::OnMessageReceived(chip::Messaging::ExchangeContext * ec,
+                                                       const chip::PayloadHeader & payloadHeader,
+                                                       chip::System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    ChipLogDetail(BDX, " %s: message " ChipLogFormatMessageType " protocol " ChipLogFormatProtocolId, __FUNCTION__,
+                  payloadHeader.GetMessageType(), ChipLogValueProtocolId(payloadHeader.GetProtocolID()));
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
```

Comment:
OK, so if err is an error and we do WillSendMessage and then pump for events.... but there aren't any events, because we errored here, what will happen?  How will things get cleaned up?

What is the exact contract here when HandleMessageReceived returns an error?  Can it have queued up events in the process?

Reply:
i just decided to abort the transfer for all errors generated by this. Since this is synchronous BDX we should get either a request or a response thats known. if the msg received is null or if the BDX message type is unknown that would indicate an issue with the session itself. at least thats what i thought. Please let me know your thoughts.